File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        boolean[][] visited = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (searchWord(board, word, 0, i, j, visited)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean searchWord(char[][] board, String word, int index, int row, int col, boolean[][] visited) {
        if (index == word.length()) {
            return true; 
        }

        if (row < 0 || row >= board.length || col < 0 || col >= board[0].length || visited[row][col] || board[row][col] != word.charAt(index)) {
            return false;
        }

        visited[row][col] = true;

        boolean result = searchWord(board, word, index + 1, row + 1, col, visited)
                || searchWord(board, word, index + 1, row - 1, col, visited)
                || searchWord(board, word, index + 1, row, col + 1, visited)
                || searchWord(board, word, index + 1, row, col - 1, visited);

        visited[row][col] = false;
        return result;
    }
}

File: 79-word-search/word-search.c
bool searchWord(char** board, char* word, int index, int row, int col, bool** visited, int m, int n) {
    if (index == strlen(word)) {
        return true;
    }

    if (row < 0 || row >= m || col < 0 || col >= n || visited[row][col] || board[row][col] != word[index]) {
        return false;
    }

    visited[row][col] = true;

    bool result = searchWord(board, word, index + 1, row + 1, col, visited, m, n)
        || searchWord(board, word, index + 1, row - 1, col, visited, m, n)
        || searchWord(board, word, index + 1, row, col + 1, visited, m, n)
        || searchWord(board, word, index + 1, row, col - 1, visited, m, n);

    visited[row][col] = false;

    return result;
}

bool exist(char** board, int boardSize, int* boardColSize, char* word) {
    bool** visited = (bool**)malloc(boardSize * sizeof(bool*));
    for (int i = 0; i < boardSize; i++) {
        visited[i] = (bool*)malloc(boardColSize[i] * sizeof(bool));
        memset(visited[i], false, boardColSize[i] * sizeof(bool));
    }
    for (int i = 0; i < boardSize; i++) {
        for (int j = 0; j < boardColSize[i]; j++) {
            if (searchWord(board, word, 0, i, j, visited, boardSize, boardColSize[i])) {
                for (int k = 0; k < boardSize; k++) {
                    free(visited[k]);
                }
                free(visited);
                return true;
            }
        }
    }

    for (int i = 0; i < boardSize; i++) {
        free(visited[i]);
    }
    free(visited);
    return false;
}